### ISSUE TITLE
Linking errors with GCC 10

### ISSUE THREAD

**Issue by FOTMarut** [FWGS/xash3d-fwgs#161]
### Problem description
Building with GCC 10 results in several linking errors in ref_gl and ref_soft.

The reason for this is that GCC 10 switched default behaviour from `-fcommon` to `-fno-common`, which improves performance with global variable references but makes it impossible for the linker to merge tentative definitions of the same symbol in different translation units.

In layman's terms, it's not ok anymore if you put `int banana;` in two different C sources that you compile and link together, you've got to make one of them `extern`. It's never been standard-compliant anyway.

This is a problem with the API function pointer declarations inside `ref_gl/gl_export.h` and a few other globals.

### Proposed solutions
The lousy solution is just to add `-fcommon` to the parameters for ref_gl and ref_soft.
A more proper one is to actually fix the problem and remove duplicate definitions. I have prepared a working solution, will post a follow-up soon.

**Comment by nekonomicon** [comment]
It would be better to fix duplicated variable definitions, not use -fcommon, because It's root of evil.

**Comment by FOTMarut** [comment]
Namely, the incriminated symbols are, in ref_soft:
```
qboolean insubmodel;
ref_instance_t RI;
cvar_t *r_drawentities;
cvar_t *vid_gamma;
```
plus all the symbols declared with GL_FUNCTION in `ref_gl/gl_export.h`.

For the former, I just needed to make some of them extern.
For the latter, I crafted a more involved solution:
> ref_gl/gl_export.h
```
#ifndef APIENTRY_LINKAGE
#define APIENTRY_LINKAGE extern
#endif

...

// helper opengl functions
APIENTRY_LINKAGE GLenum GL_FUNCTION( glGetError )(void);
APIENTRY_LINKAGE const GLubyte * GL_FUNCTION( glGetString )(GLenum name);

// base gl functions
APIENTRY_LINKAGE void GL_FUNCTION( glAccum )(GLenum op, GLfloat value);
APIENTRY_LINKAGE void GL_FUNCTION( glAlphaFunc )(GLenum func, GLclampf ref);
...
```

> ref_gl/gl_context.c
```
#define APIENTRY_LINKAGE
#include "gl_local.h"
#include "gl_export.h"
```

> ref_soft/r_glblit.c
```
#define APIENTRY_LINKAGE static
#include "../ref_gl/gl_export.h"
```

This way function (pointers) are declared as:

- implicitly `auto` in `ref_gl/gl_context.c`
- `extern` in all other translation units in ref_gl
- `static` in `ref_soft/r_glblit.c`, because why not. No other source file in ref_soft needs those anyway.

If you like this solution I can make a pull request. Let me know of any desired changes beforehand, please, so we don't litter the commit history.

**Comment by a1batross** [comment]
Yeah, that's what I've been afraid for A LONG time and that's why I've added -fvisibility=hidden by default, otherwise engine's variables will mess up with HLSDK, no matter if they are same kind or not.

Well, with `-fno-common` it seems to be a lot more easier. I wanted to make `gl_export.h` a standard compliant in the same way you suggested, but in the end, I just forgot about it. %)

Yeah, I'm fine with this solution.

ref_soft code also must be patched to make common symbols defined once and declared in header like r_local.h. It's port of Quake's software renderer and even if I respect old id Software, they didn't care about standards and UBs.

**Comment by FOTMarut** [comment]
Alright, cool, I'll prepare the PR.
Also, if you have a better idea for naming let me know, I'm still not sure if I'm happy with APIENTRY_LINKAGE.

**Comment by a1batross** [comment]
That's fine. It's always better to have more configurability. Yes, we have just APIENTRY but it's reserved for calling convention.

**Comment by a1batross** [comment]
BTW, after your PR will be merged, I think it should enabled by default for GCC.

It [is supported](https://gcc.gnu.org/onlinedocs/gcc-3.4.6/gcc/Code-Gen-Options.html#Code-Gen-Options) for oldest GCC version we're using, which is GCC 3.4 targetting Motorola's Linux devices.

Clang meanwhile doesn't mention it in the documentation.

**Comment by FOTMarut** [comment]
Cross-referencing #162
